src/preprocess/filter.ts

```diff
@@ -11,7 +11,16 @@ export function filter(keys: string[], input: any, filename: string) {
   const output = _.pick(input, ['$imports', '$defs', ...keys]);
   visitor.visitNode(output, 'Root', env);
   if(input.$imports) {
-    output.$imports = _.pick(input.$imports, visitor.variables);
+    // Imports are not tested and nested lookups depend on the object that is imported, for example:
+    //
+    //   $imports:
+    //     vars: vars.yaml
+    //     outputs: cfn:output:my-stack
+    //   foo: vars.foo
+    //   var: outputs.bar
+    //
+    // should preserve `vars` and `outputs`
+    output.$imports = _.pick(input.$imports, _.map(visitor.variables, (v: string) => v.split('.')[0]));
```

Comment:
I'm not not clear on what exactly `visitor` is - would `visitor.variables` be an array containing `'$imports.vars', '$imports.outputs', and so those wouldn't have matched against the `$imports` key that `_.pick` found?

Reply:
`vistor.variables` is a list of variables that are used in the body of the file. For example, for
```
foo: !$ bar
```
`vistor.variables` would be `["bar"]`.

It is used to filter out unused `$imports` for some commands, such as `template-approval`, which only cares about specific stack args fields.

https://github.com/unbounce/iidy/blob/11d4ea83167c20202a312763f263e40868b5380a/src/cfn/approval/index.ts#L24-L28

Without this, all imports must be satisfied, even if they aren't used (which is a frustration for things like environment variables which you would need to provide with a dummy value).